exes/mqtt-bridge/src/spark_plug_interface.cpp

```diff
@@ -0,0 +1,342 @@
+
+namespace asio = boost::asio;
+
+template <class config_t, class mqtt_client_t>
+spark_plug_interface<config_t, mqtt_client_t>::spark_plug_interface(asio::io_context& io_ctx)
+    : io_ctx_(io_ctx), ncmd_topic_(topic_formatter({ constants::namespace_element, config_.value().group_id, constants::ncmd,
+                                                     config_.value().node_id })),
+      mqtt_will_topic_(topic_formatter(
+          { constants::namespace_element, config_.value().group_id, constants::ndeath, config_.value().node_id })),
+      ndata_topic_(topic_formatter(
+          { constants::namespace_element, config_.value().group_id, constants::ndata, config_.value().node_id })) {
+  const std::string_view mqtt_will_payload{ make_will_payload() };
+
+  mqtt_client_ = std::make_unique<mqtt_client_t>(io_ctx_, mqtt_will_topic_, mqtt_will_payload);
+}
+
+template <class config_t, class mqtt_client_t>
+auto spark_plug_interface<config_t, mqtt_client_t>::make_will_payload() -> std::string {
+  /// Construct MQTT will payload
+  Payload payload;
+  payload.set_timestamp(timestamp_milliseconds().count());
+
+  auto* metric = payload.add_metrics();
+
+  metric->set_name("bdSeq");
+  metric->set_timestamp(timestamp_milliseconds().count());
+  metric->set_datatype(8);
+  metric->set_long_value(0);
+
+  std::string payload_string;
+  payload.SerializeToString(&payload_string);
+
+  return payload_string;
+}
+
+/// Create a timestamp in milliseconds from 1 Jan 1970 according to spec
+template <class config_t, class mqtt_client_t>
+auto spark_plug_interface<config_t, mqtt_client_t>::timestamp_milliseconds() -> std::chrono::milliseconds {
+  using std::chrono::duration_cast;
+  using std::chrono::milliseconds;
+  using std::chrono::system_clock;
+  return duration_cast<milliseconds>(system_clock::now().time_since_epoch());
+}
+
+template <class config_t, class mqtt_client_t>
+auto spark_plug_interface<config_t, mqtt_client_t>::set_current_values(std::vector<structs::signal_data> const& metrics)
+    -> void {
+  nbirth_ = metrics;
+}
+
+template <class config_t, class mqtt_client_t>
+auto spark_plug_interface<config_t, mqtt_client_t>::send_current_values() -> void {
+  if (mqtt_client_) {
+    Payload payload;
+    payload.set_timestamp(timestamp_milliseconds().count());
+    seq_ = 0;
+    payload.set_seq(seq_);
+
+    auto* node_rebirth = payload.add_metrics();
+    node_rebirth->set_name(constants::rebirth_metric.data());
+    node_rebirth->set_timestamp(timestamp_milliseconds().count());
+    node_rebirth->set_datatype(11);
+    node_rebirth->set_boolean_value(false);
+    node_rebirth->set_is_transient(false);
+    node_rebirth->set_is_historical(false);
+    node_rebirth->set_is_null(false);
+
+    auto* bd_seq_metric = payload.add_metrics();
+
+    bd_seq_metric->set_name("bdSeq");
+    bd_seq_metric->set_timestamp(timestamp_milliseconds().count());
+    bd_seq_metric->set_datatype(8);
+    bd_seq_metric->set_long_value(0);
+
```

Comment:
Þessi tala þarf að breytast við hvert nbirth sýnist mér.
Sjá https://stackoverflow.com/questions/55465375/what-is-the-difference-between-bdseq-and-seq-number-in-mqtt-sparkplug

Reply:
Ahh okei, bæti við member variable og falli sem heldur utan um gildið.